transports/noise/src/lib.rs

```diff
@@ -219,6 +223,59 @@ where
     }
 }
 
+impl<T> InboundSecurityUpgrade<T> for Config
+where
+    T: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+{
+    type Output = Output<T>;
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<(PeerId, Self::Output), Self::Error>>;
+
+    fn secure_inbound(self, socket: T, _: Self::Info) -> Self::Future {
+        async move {
+            let mut state = self.into_responder(socket)?;
+
+            handshake::recv_empty(&mut state).await?;
+            handshake::send_identity(&mut state).await?;
+            handshake::recv_identity(&mut state).await?;
+
+            let (pk, io) = state.finish()?;
+
+            let expected = pk.to_peer_id();
+            Ok((expected, io))
+        }
+        .boxed()
+    }
+}
+
+impl<T> OutboundSecurityUpgrade<T> for Config
+where
+    T: AsyncRead + AsyncWrite + Unpin + Send + 'static,
+{
+    type Output = Output<T>;
+    type Error = Error;
+    type Future = BoxFuture<'static, Result<(PeerId, Self::Output), Self::Error>>;
+
+    fn secure_outbound(self, socket: T, _: Self::Info, peer_id: Option<PeerId>) -> Self::Future {
```

Comment:
Eventually, we will remove the other implementation. Thus, I think to reduce duplication, we can implement `InboundConnectionUpgrade` by delegating to `InboundSecurityUpgrade` and equivalent for outbound.

Reply:
Already done.